src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -499,32 +499,48 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
-                int printNameNameOffset = sizeof(Interop.Kernel32.REPARSE_DATA_BUFFER) + rdb.ReparseBufferSymbolicLink.PrintNameOffset;
-                int printNameNameLength = rdb.ReparseBufferSymbolicLink.PrintNameLength;
+                // PrintName can ONLY return a NT path if the link was created explicitly targeting a file/folder in such way.
+                //   e.g: mklink /D linkName \??\C:\path\to\target.
+
+                if (rbSymlink.ReparseTag == Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK)
+                {
+                    int printNameOffset = sizeof(Interop.Kernel32.SymbolicLinkReparseBuffer) + rbSymlink.PrintNameOffset;
+                    int printNameLength = rbSymlink.PrintNameLength;
 
-                Span<char> targetPath = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(printNameNameOffset, printNameNameLength));
-                Debug.Assert((rdb.ReparseBufferSymbolicLink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) == 0 || !PathInternal.IsExtended(targetPath));
+                    Span<char> targetPathSymlink = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(printNameOffset, printNameLength));
+                    Debug.Assert((rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) == 0 || !PathInternal.IsExtended(targetPathSymlink));
 
-                if (returnFullPath && (rdb.ReparseBufferSymbolicLink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0)
+                    if (returnFullPath && (rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0)
+                    {
+                        // Target path is relative and is for ResolveLinkTarget(), we need to append the link directory.
+                        return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPathSymlink);
+                    }
+
+                    return targetPathSymlink.ToString();
+                }
+                else if (rbSymlink.ReparseTag == Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_MOUNT_POINT)
                 {
-                    // Target path is relative and is for ResolveLinkTarget(), we need to append the link directory.
-                    return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPath);
+                    success = MemoryMarshal.TryRead(bufferSpan, out Interop.Kernel32.MountPointReparseBuffer rbMountPoint);
+                    Debug.Assert(success);
+
+                    int printNameOffset = sizeof(Interop.Kernel32.MountPointReparseBuffer) + rbMountPoint.PrintNameOffset;
+                    int printNameLength = rbMountPoint.PrintNameLength;
+
+                    Span<char> targetPathMountPoint = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(printNameOffset, printNameLength));
+
+                    // Unlink symlinks, mount point paths cannot be relative
+                    Debug.Assert(!PathInternal.IsExtended(targetPathMountPoint));
```

Comment:
This assert doesn't verify that the path is not relative. I think you should use this:
```suggestion
                    Debug.Assert(!PathInternal.IsExtended(targetPathMountPoint));
                    Debug.Assert(Path.IsPathFullyQualified(targetPathMountPoint));
```